chainerx_cc/chainerx/native/native_device/linalg.cc

```diff
@@ -0,0 +1,188 @@
+    sgesv_(&n, &nrhs, a, &lda, ipiv, b, &ldb, info);
+}
+
+template <>
+void Getrf<double>(int m, int n, double* a, int lda, int* ipiv, int* info) {
+    dgetrf_(&m, &n, a, &lda, ipiv, info);
+}
+
+template <>
+void Getrf<float>(int m, int n, float* a, int lda, int* ipiv, int* info) {
+    sgetrf_(&m, &n, a, &lda, ipiv, info);
+}
+
+template <>
+void Getri<double>(int n, double* a, int lda, int* ipiv, double* work, int lwork, int* info) {
+    dgetri_(&n, a, &lda, ipiv, work, &lwork, info);
+}
+
+template <>
+void Getri<float>(int n, float* a, int lda, int* ipiv, float* work, int lwork, int* info) {
+    sgetri_(&n, a, &lda, ipiv, work, &lwork, info);
+}
+#endif
+
+template <typename T>
+void SolveImpl(const Array& a, const Array& b, const Array& out) {
+    Device& device = a.device();
+    Dtype dtype = a.dtype();
+
+    Array lu_matrix = Empty(a.shape(), dtype, device);
+    device.backend().CallKernel<CopyKernel>(a.Transpose(), lu_matrix);
+    T* lu_ptr = static_cast<T*>(internal::GetRawOffsetData(lu_matrix));
+
+    int n = a.shape()[0];
+    int nrhs = 1;
+    if (b.ndim() == 2) {
+        nrhs = b.shape()[1];
+    }
+
+    Array ipiv = Empty(Shape{n}, Dtype::kInt32, device);
+    int* ipiv_ptr = static_cast<int*>(internal::GetRawOffsetData(ipiv));
+
+    device.backend().CallKernel<CopyKernel>(b, out);
+    T* out_ptr = static_cast<T*>(internal::GetRawOffsetData(out));
+
+    int info;
+    Gesv(n, nrhs, lu_ptr, n, ipiv_ptr, out_ptr, n, &info);
+
+    if (info != 0) {
+        throw ChainerxError{"Unsuccessfull gesv (Solve) execution. Info = ", info};
+    }
+}
+
+template <typename T>
+void InverseImpl(const Array& a, const Array& out) {
+    Device& device = a.device();
+    Dtype dtype = a.dtype();
+
+    device.backend().CallKernel<CopyKernel>(a, out);
+    T* out_ptr = static_cast<T*>(internal::GetRawOffsetData(out));
+
+    int n = a.shape()[0];
+
+    Array ipiv = Empty(Shape{n}, Dtype::kInt32, device);
+    int* ipiv_ptr = static_cast<int*>(internal::GetRawOffsetData(ipiv));
+
+    int info;
+    Getrf(n, n, out_ptr, n, ipiv_ptr, &info);
+    if (info != 0) {
+        throw ChainerxError{"Unsuccessfull getrf (LU) execution. Info = ", info};
+    }
+
+    int buffersize = -1;
+    T work_size;
+    Getri(n, out_ptr, n, ipiv_ptr, &work_size, buffersize, &info);
+    buffersize = static_cast<int>(work_size);
+    Array work = Empty(Shape{buffersize}, dtype, device);
+    T* work_ptr = static_cast<T*>(internal::GetRawOffsetData(work));
+
+    Getri(n, out_ptr, n, ipiv_ptr, work_ptr, buffersize, &info);
+    if (info != 0) {
+        throw ChainerxError{"Unsuccessfull getri (Inverse LU) execution. Info = ", info};
+    }
+}
+
+}  // namespace
+
+class NativeSolveKernel : public SolveKernel {
+public:
+    void Call(const Array& a, const Array& b, const Array& out) override {
+#ifndef CHAINERX_ENABLE_LAPACK
+        throw ChainerxError{"LAPACK is not linked to ChainerX."};
```

Comment:
You'll get unused variable warnings here when LAPACK is not available. You can do the following (we do this in other places).
```c++
(void)a;  // unused
(void)b;  // unused
(void)out;  // unused
throw ChainerxError{"LAPACK is not linked to ChainerX."};
```